### ISSUE TITLE
bug(StreamHashJoin): redundant/unexpected output ops for update

### ISSUE THREAD

**Issue by xiangjinwu** [risingwavelabs/risingwave#14835]
### Describe the bug

Refer to repro steps and expected behavior.

### To Reproduce

Add `tracing::warn!("proj-input: {}", chunk.to_pretty_with_schema(&sch));` in phase 1 `Message::Chunk`.
Also `let sch = input.schema().clone();` before `inner.execute()`.
(Are there simpler ways to inspect on input/output chunks of a specific executor?)
https://github.com/risingwavelabs/risingwave/blob/7e937c22fcacf58fc8c7b9aca990862d5f55ac12/src/stream/src/executor/project.rs#L182-L189

```
create table posts (pid int primary key, title varchar);
create table comments (cid int primary key, pid int);
create materialized view num_comments as select pid, count(*) as num from comments group by pid;
create materialized view stat_by_posts as select p.pid, title, coalesce(num, 0) from posts p left join num_comments n on p.pid = n.pid;
insert into posts values (1, 'hello');
insert into comments values (11, 1);
insert into comments values (12, 1);
```

The input chunk of project executor, which is the output chunk of join executor, appears as following in the added log:
```
+----+-----------+-------------+------------------+------------------+
|    | posts.pid | posts.title | num_comments.num | num_comments.pid |
+----+-----------+-------------+------------------+------------------+
| U- | 1         | hello       | 1                | 1                |
| U+ | 1         | hello       |                  |                  |
| U- | 1         | hello       |                  |                  |
| U+ | 1         | hello       | 2                | 1                |
+----+-----------+-------------+------------------+------------------+
```

Note the above uses a left outer join. If it was inner join, the chunks logged would be:
```
+---+-----------+-------------+------------------+------------------+
|   | posts.pid | posts.title | num_comments.num | num_comments.pid |
+---+-----------+-------------+------------------+------------------+
| - | 1         | hello       | 1                | 1                |
| + | 1         | hello       | 2                | 1                |
+---+-----------+-------------+------------------+------------------+
```

### Expected behavior

By intuition, the ideal output of join executor in this case is (regardless of left outer or inner):
```
+----+-----------+-------------+------------------+------------------+
|    | posts.pid | posts.title | num_comments.num | num_comments.pid |
+----+-----------+-------------+------------------+------------------+
| U- | 1         | hello       | 1                | 1                |
| U+ | 1         | hello       | 2                | 1                |
+----+-----------+-------------+------------------+------------------+
```

Semantically, they look equivalent and we may think it has little impact on downstream processing. However, when sinking with upsert behavior, it results in unnecessary deletion where a single insert in sufficient.

### How did you deploy RisingWave?

./risedev playground

### The version of RisingWave

main (f16a62386e2edd7b3accd85d9b31dcc6eda307ac)

### Additional context

Derived from a customer reported issue about upsert sink containing unexpected tombstones (i.e. deletions), even when derived upstream key matches with specified downstream key. cc @tabVersion @st1page

**Comment by yuhao-su** [comment]
It is a legitimate output for the current join implementation. Doing such optimization can be complicated, not to mention we can't avoid the "redundant" U- U+ mentioned above if the two input are in different chunks. 

If it really frustrating to have unnecessary delete to sink downstream, we can consider adding a compaction before sink.

**Comment by BugenZhao** [comment]
> Are there simpler ways to inspect on input/output chunks of a specific executor?

Yes, but currently not for specific executor.

https://github.com/risingwavelabs/risingwave/pull/12024#issuecomment-1706002065

**Comment by fuyufjh** [comment]
> we can consider adding a compaction before sink.

Done in https://github.com/risingwavelabs/risingwave/pull/15345

However, personally, I still hope to complete this issue. #15345 is somehow a "patch" instead of a native solution.

**Comment by yuhao-su** [comment]
should be resolved in #17450
